File: 150 HackerRank.java
// 100% HackerRank

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;
import java.util.*;
import java.util.PriorityQueue;
import java.util.Collections;
import java.util.ArrayList;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        
        int[][] arr = new int[N][];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            arr[i] = new int[i + 1];
            for (int j = 0; j < i + 1; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        
        int[][] partialSum = new int[N][];
        for (int i = 0; i < N; i++) {
            partialSum[i] = new int[i + 1];
            partialSum[i][0] = arr[i][0];
            for (int j = 1; j < i + 1; j++) {
                partialSum[i][j] = partialSum[i][j - 1] + arr[i][j];
            }
        }
        
        PriorityQueue<Integer> heap = new PriorityQueue<>(Collections.reverseOrder());
        
        for (int r = 0; r < N; r++) {
            for (int c = 0; c <= r; c++) {
                int currentSum = 0;
                int maxSize = N - r;
                for (int s = 1; s <= maxSize; s++) {
                    int rowIndex = r + s - 1;
                    int colStart = c;
                    int colEnd = c + s - 1;
                    
                    int rowSegment = partialSum[rowIndex][colEnd];
                    if (colStart > 0) {
                        rowSegment -= partialSum[rowIndex][colStart - 1];
                    }
                    
                    currentSum += rowSegment;
                    
                    if (heap.size() < K) {
                        heap.offer(currentSum);
                    } else if (currentSum < heap.peek()) {
                        heap.poll();
                        heap.offer(currentSum);
                    }
                }
            }
        }
        
        ArrayList<Integer> result = new ArrayList<>();
        while (!heap.isEmpty()) {
            result.add(heap.poll());
        }
        Collections.sort(result);
        
        for (int sum : result) {
            System.out.println(sum);
        }
    }
}

File: 150 HackerRank.py
# 150% HackerRank

import heapq

N , K = map(int, input().split())

arr = []
index = 2
for i in range(N):
    row = list(map(int, input().split()))
    arr.append(row)
    index += (i + 1)

partial_sum = []
for i in range(N):
    row_psum = [0] * (i + 1)
    row_psum[0] = arr[i][0]
    for j in range(1, i + 1):
        row_psum[j] = row_psum[j - 1] + arr[i][j]
    partial_sum.append(row_psum)

heap = []

for r in range(N):
    for c in range(r + 1):
        current_sum = 0
        max_size = N - r
        for s in range(1, max_size + 1):
            row_idx = r + s - 1
            col_start = c
            col_end   = c + (s - 1)

            row_segment = partial_sum[row_idx][col_end]
            if col_start > 0:
                row_segment -= partial_sum[row_idx][col_start - 1]

            current_sum += row_segment

            if len(heap) < K:
                heapq.heappush(heap, -current_sum)  # store negative
            else:
                if -current_sum > heap[0]:
                    heapq.heapreplace(heap, -current_sum)

result = [-val for val in heap]
result.sort()

for val in result:
    print(val)

